akka-pem-reader/src/main/scala/akka/remote/artery/tcp/PEMConfigSSLEngineProvider.scala

```diff
@@ -0,0 +1,99 @@
+package akka.remote.artery.tcp
+
+import java.io.FileInputStream
+import java.io.FileNotFoundException
+import java.io.IOException
+import java.security.GeneralSecurityException
+import java.security.KeyStore
+import java.security.cert.Certificate
+import java.security.cert.CertificateFactory
+
+import akka.actor.ActorSystem
+import akka.pki.pem.PEMPrivateKeyLoader
+import javax.net.ssl.KeyManager
+import javax.net.ssl.KeyManagerFactory
+import javax.net.ssl.TrustManager
+import javax.net.ssl.TrustManagerFactory
+
+class PEMConfigSSLEngineProvider(system: ActorSystem) extends ConfigSSLEngineProvider(system) {
```

Comment:
This approach won't scale. Extending `ConfigSSLEngineProvider` and then overriding `keyManagers` and `trustManagers` feels wrong.

Instead, I'm playing with the idea the an `SSLEngineProvider` is the composition of the following:

```
keyManagersProvider: (Config) => Array[KeyManager]
trustManagersProvider: (Config) => Array[TrustManager]
secureRandomProvider: (Config) => SecureRandom

sslContextFactory: (Array[KeyManager], Array[TrustManager], SecureRandom) => SSLContext
sslEngineFactory: (SSLContext, Config) => SSLEngine

// the following would be used in `verifyClientSession` and `verifyServerSession`, 
// e.g. to authorize the remote 
sslSessionVerifiers: Array[SSLSessionVerifiers]
```

With the above in mind, instead of building custom use cases via inheritance as I did here, users coudl compose more easily.

Reply:
Note to self: add a `reference.conf`

Reply:
it will be very rare to implement custom `SSLEngineProvider` so we shouldn't over-engineer with too much api and flexibility

Reply:
extracting common code to shared internals might be doable, especially if you move this alongside the other in akka-remote.

Reply:
that said, I'm not against a few more extension points but it must not turn into a new SslConfig. We deliberately kept things rather raw for SSLEngineProvider because more important to use well known standard Java api for these things even though it's ux is terrible.

Reply:
> it must not turn into a new SslConfig

I agree with that 120%